ssdp.md

```diff
@@ -0,0 +1,135 @@
+* Open questions:
+    * how to send the endpoint of the receiver service: one solution is to use a HTTP header parameter in the response of the HTTP GET request for device description. DIAL uses this solution to send the endpoint of the DIAL server in the `Application-URL` HTTP response header. Another solution is to extend the XML device description with a new element to define the endpoint.
+    * How to check if the display can open a certain URL or not. One solution is to extend the XML device description with new elements to allow a display to express its capabilities and the controller can do the check. Another possible solution is to ask the receiver service using the provided endpoint by sending the presentation URLs. 
+
+**Method 2**: This method uses only the SSDP messages without requesting the device description XML. Idea is to send the presentation URLs in a new header of the SSDP search message. Only receivers that can open at least one of the URLs response to the search request. The search response contains also a new header for device friendly name and another header for service endpoint (section 1.1.3 of the UPnP device architecture document allows to use vendor specific headers). The search response can still send the device description URL in the Location header to stay compatible with UPnP but controller devices don’t need to use it since all necessary information are send directly in the SSDP search response. This method is more efficient since no additional HTTP calls and XML parsing is needed. Below are the steps that illustrate the idea of this method:
+
+* New display with presentation receiver service that appears on the network advertises the following SSDP alive message. The new header `FRIENDLY-NAME.example.org` is used for friendly name and `PRESENTATION-ENDPOINT.example.org` for endpoint of the receiver service. The value `urn:example-org:service:osp:1` of header `NT` is used for presentation receiver service.
+
+    ```
+    NOTIFY * HTTP/1.1
+    HOST: 239.255.255.250:1900
+    CACHE-CONTROL: max-age = seconds until advertisement expires
+    LOCATION: URL of the device description
+    NTS: ssdp:alive
+    SERVER: OS/version UPnP/1.0 product/version
+    USN: advertisement UUID
+    NT: urn:example-org:service:osp:1
+    FRIENDLY-NAME.example.org: My Presentation Display
+    PRESENTATION-ENDPOINT.example.org: 192.168.1.100:3000
+    ```
+
+* A controller sends the following SSDP search message to the multicast address. The new header `PRESENTATION-URLS.example.org` allows the controller to send the presentation URLs to the display. 
+
+    ```
+    M-SEARCH * HTTP/1.1
+    HOST: 239.255.255.250:1900
+    MAN: "ssdp:discover"
+    MX: seconds to delay response
+    ST: urn:example-org:service:osp:1
+    PRESENTATION-URLS.example.org: http://example.com/foo.html, http://example.com/bar.html
```

Comment:
I think if there's a privacy preserving way to optimize away some connections to the receiver service that would be a benefit.   If the primary benefit is to separate out generic endpoints (https://*)  from application-specific ones (e.g., casts://*) a simple scheme advertisement approach might work.